### ISSUE TITLE
Unexpected parsing error in shorthand lambda

### ISSUE THREAD

**Issue by DedSec256** [dotnet/fsharp#16220]
```fsharp
let _ = 1 |> _.ToString()
1 |> _.ToString()
```

![image](https://github.com/dotnet/fsharp/assets/26364714/5f04a0fc-0a4d-484a-b419-46eb423b82bc)

.NET 8.0.100-rc.2.23502.2

**Comment by abonie** [comment]
Can you share sdk version?

**Comment by DedSec256** [comment]
@abonie, reproducible in .NET 8.0.100-rc.2.23502.2
Also in can be reproduced via test in the latest main branch

![image](https://github.com/dotnet/fsharp/assets/26364714/3015ecb0-ec99-42a8-979c-227e7e51cdda)

```fsharp
[<Fact>]
let Test () =
    Fsx """
let _ = 1 |> _.ToString()
1 |> _.ToString()"""
    |> withLangVersion80
    |> typecheck
    |> shouldSucceed
```

**Comment by abonie** [comment]
Hmm... curious thing is that it does not repro for me in fsi or when I create a project with `dotnet new console ...` but it does repro when I create a new console app project with VS or with the unit test method @DedSec256 proposed.

**Comment by vzarytovskii** [comment]
Might be because it's on a top level, due to how we codegen/lift stuff? Is there a difference top level vs inside function/method?

**Comment by auduchinok** [comment]
> Might be because it's on a top level, due to how we codegen/lift stuff? Is there a difference top level vs inside function/method?

Yes, there're some parsing differences IIRC, i.e. not every expression normally can be parsed at top level. However, here the lambda is inside a binary application expression, so I'd probably expect it to work (though, it depends on the implementation).

**Comment by abonie** [comment]
Yeah, if the code snippet is not top level it does not repro anymore

**Comment by T-Gro** [comment]
Yep, this is it.
This error message is done based on the shape of the parsed SyntaxTree - this is the error I added to make things like _."a" invalid, but the top-level statement situation needs special casing here.

**Comment by abonie** [comment]
Although if I create, build and run a project with `dotnet` CLI tool, it is still top-level but it does not repro. What's up with that?

**Comment by T-Gro** [comment]
IMO the fact that "recover" can also be EOF is what causes this.

**Comment by abonie** [comment]
Yeah, you are right @T-Gro , behavior is the same whether I create a project with dotnet CLI or VS, it was about whether I pasted this snippet above or below the printfn that was already there from the template.

**Comment by auduchinok** [comment]
This was fixed in #16240.
